### PR TITLE
16: Routing I - Login (Part 3)

### PR DESCRIPTION
**Pull request by gabrielizalo**
# Summary

* Third task of Routing: Redux form del login
   API Integration.

# Trello Card

https://trello.com/c/uRUucKHu/9-routing-i-login

# Screenshot

gabo s tic tac toe

### CODE REVIEW

File: react/react-training/.env [hunk 1]
```diff
@@ -0,0 +1 @@
+REACT_APP_API_BASE_URL=http://localhost:3001/api/
```
Comment (wfolini) [inline react/react-training/.env:1]: This file contains sensible info, it should not be commited. Add it to .gitignore file

File: react/react-training/src/app/components/AppRoutes/components/AuthRoute/index.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import React, { Component } from 'react';
+import { Redirect, Route } from 'react-router';
+import PropTypes from 'prop-types';
+
+export default class AuthComponent extends Component {
```
Comment (wfolini) [inline react/react-training/src/app/components/AppRoutes/components/AuthRoute/index.js]: `export default ..` always to the bottom of the file. Also, this comp could be a functional comp

File: react/react-training/src/app/components/AppRoutes/index.js [hunk 3]
```diff
@@ -1,19 +1,39 @@
...
+import AuthRoute from './components/AuthRoute';
+
+class AppRoutes extends Component {
   render() {
+    const { userIsLogged } = this.props;
     return (
       <Router>
         <Switch>
-          <Route exact path={ROUTES.HOME} component={Login} />
+          <AuthRoute exact path={ROUTES.HOME} component={Login} userIsLogged={userIsLogged} />
```
Comment (wfolini) [inline react/react-training/src/app/components/AppRoutes/index.js:18]: What would happen if I'm not logged in but I want to go to the path `/home` entering through the navigation bar? I think it allows me and it should not happen.

File: react/react-training/src/app/screens/Login/components/CustomField/index.js [hunk 4]
```diff
@@ -1,16 +1,18 @@
...
 
-export function CustomField(props) {
-  const { input, type, className, placeholder, meta } = props;
-  return (
-    <div className={styles.input}>
-      <input {...input} type={type} className={className} placeholder={placeholder} />
-      {meta.error && meta.touched && <div className={styles.error}>{meta.error}</div>}
-    </div>
-  );
+export class CustomField extends Component {
```
Comment (wfolini) [inline react/react-training/src/app/screens/Login/components/CustomField/index.js]: This component as a function was right
```js
function CustomField({ input, type, className, placeholder, meta }) {
```

File: react/react-training/src/redux/login/actions.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  login: 'LOGIN',
+  loginSuccess: 'LOGIN_SUCCESS',
+  loginFailure: 'LOGIN_FAILURE'
+};
+
+export const actionsCreators = {
+  login: (username, password) => async dispatch => {
+    dispatch({ type: actionTypes.login });
+    const response = await UserService.login(username, password);
+    if (response && response.data) {
```
Comment (wfolini) [inline react/react-training/src/redux/login/actions.js]: with asking for `(response.ok)` is enough. I'd delete this if

File: react/react-training/src/redux/login/actions.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      if (response.ok) {
+        const mySession = response.data.id ? response.data.id : MSG_UNKNOWN_ID;
```
Comment (wfolini) [inline react/react-training/src/redux/login/actions.js]: `response.data.id || MSG_UNKNOWN_ID`

File: react/react-training/src/redux/login/actions.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import UserService from '@services/UserService.js';
+import { MSG_UNKNOWN_ID, ERROR_READING_RESPONSE, ERROR_LOGIN_SERVICE } from '@screens/Login/validation';
+
+export const actionTypes = {
+  login: 'LOGIN',
+  loginSuccess: 'LOGIN_SUCCESS',
+  loginFailure: 'LOGIN_FAILURE'
```
Comment (wfolini) [inline react/react-training/src/redux/login/actions.js]: As this actions are const, they must be written in upper case
```
  LOGIN: 'LOGIN',
  LOGIN_SUCCESS: 'LOGIN_SUCCESS',
  LOGIN_FAILURE: 'LOGIN_FAILURE'
```

File: react/react-training/src/redux/login/actions.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+export const actionsCreators = {
+  login: (username, password) => async dispatch => {
+    dispatch({ type: actionTypes.login });
+    const response = await UserService.login(username, password);
+    if (response && response.data) {
+      if (response.ok) {
+        const mySession = response.data.id ? response.data.id : MSG_UNKNOWN_ID;
+        dispatch({
+          type: actionTypes.loginSuccess,
+          userSession: mySession
```
Comment (wfolini) [inline react/react-training/src/redux/login/actions.js]: replace by `payload`
```js
payload: mySession
```

File: react/react-training/src/redux/login/actions.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        });
+      } else {
+        const myError =
+          response.data.error && response.data.error.message && response.data.error.statusCode
+            ? `Error ${response.data.error.statusCode} - ${response.data.error.message}`
+            : ERROR_READING_RESPONSE;
+        dispatch({
+          type: actionTypes.loginFailure,
+          userLoginError: myError
```
Comment (wfolini) [inline react/react-training/src/redux/login/actions.js]: same about `payload`

File: react/react-training/src/redux/login/actions.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        });
+      }
+    } else {
+      dispatch({
+        type: actionTypes.loginFailure,
+        userLoginError: ERROR_LOGIN_SERVICE
```
Comment (wfolini) [inline react/react-training/src/redux/login/actions.js]: same about `payload`

**Review by wfolini — CHANGES_REQUESTED** [review]
Great! Some comments!

File: react/react-training/src/redux/login/actions.js (hunk 5 above)
Comment (gabrielizalo) [reply react/react-training/src/redux/login/actions.js]: Hola @WFolini, te pregunto…
Me dices que puedo eliminar la linea `if (response && response.data) {`
Sin embargo es necesaria.. Si el servicio está fuera de línea no hay `response` y por tanto la app saca este error:
gabo s tic tac toe 2018-11-15 16-08-28
Con ella valido si hay servicio que responda… ¿O cómo puedo validar esto?

File: react/react-training/src/app/components/AppRoutes/index.js (hunk 3 above)
Comment (gabrielizalo) [reply react/react-training/src/app/components/AppRoutes/index.js:18]: Hola @WFolini 
Si el usuario ingresa a la url /home no sale nada de la app como puedes ver en la siguiente imagen. ¿Qué debería hacer?
gabo s tic tac toe - home

**Comment by gabrielizalo** [comment]
We are super hyper mega ok 👍

**Review by wfolini — APPROVED** [review]
Excellent! Apply this comment in the next pr!

File: react/react-training/src/constants/routes.js [hunk 11]
```diff
@@ -1,4 +1,8 @@
 export const ROUTES = {
+  // Authentication
```
Comment (wfolini) [inline react/react-training/src/constants/routes.js:2]: This repeated routes isn't necessary. I'd delete it

**Review by wfolini — APPROVED** [review]

**Pull request merged**